src/Model/UserModel.hs

```diff
@@ -1,37 +1,60 @@
 {-# LANGUAGE StrictData #-}
 module Model.UserModel where
 
+import DB.User
+import qualified Data.Map as M
+import Data.Password.Argon2 (Password, mkPassword)
 import qualified Data.Text as T
-
-import Validation
-data NewUser =
-  NewUser { usernameParam    :: {-# UNPACK#-} Text
-          , emailParam       :: {-# UNPACK#-} Text
-          , passwordParam    :: {-# UNPACK#-} Text
-          , displayNameParam :: {-# UNPACK#-} Text
-          } deriving stock (Show, Eq)
+import Data.Time
+import Data.UUID.V4 (nextRandom)
+import Web.Form
+import Web.Sessions
 
 data NewUserValidationError
   = EmptyUsername
   | EmptyDisplayName
   | TooShortPassword
   | InvalidEmailAddress
-
-validateNewUser :: NewUser -> Either (NonEmpty NewUserValidationError) ()
-validateNewUser NewUser{..} = validationToEither $
-  validateEmptyUsername usernameParam
-    <* validateEmptyDisplayName displayNameParam
-    <* validateShortPassword passwordParam
-    <* validateInvalidEmailAddress emailParam
-
-validateEmptyUsername :: Text -> Validation (NonEmpty NewUserValidationError) ()
-validateEmptyUsername name = failureIf (T.null name) EmptyUsername
-
-validateEmptyDisplayName:: Text -> Validation (NonEmpty NewUserValidationError) ()
-validateEmptyDisplayName name = failureIf (T.null name) EmptyDisplayName
-
-validateShortPassword :: Text -> Validation (NonEmpty NewUserValidationError) ()
-validateShortPassword password = failureIf (T.length password < 8) TooShortPassword
-
-validateInvalidEmailAddress :: Text -> Validation (NonEmpty NewUserValidationError) ()
-validateInvalidEmailAddress email = failureUnless (T.isInfixOf "@" email) InvalidEmailAddress
+  | MissingField Text
+
+instance ErrorToAssign NewUserValidationError where
+  putErrorAssign err =
+    case err of
+      EmptyUsername       -> putAssign "form_error_username" "Username cannot be empty"
+      EmptyDisplayName    -> putAssign "form_error_displayname" "Display name cannot be empty"
+      TooShortPassword    -> putAssign "form_error_password" "Password cannot be smaller than 8 characters"
+      InvalidEmailAddress -> putAssign "form_error_email" "Email address is invalid"
+      MissingField field  -> putAssign ("form_error_" <> field) ("Missing required field: " <> field)
+
+validateNewUser :: [(Text,Text)] -> IO (FormValidation NewUserValidationError User)
+validateNewUser params = do
+  let paramMap = M.fromList params
+  ts <- getCurrentTime
+  userId <- UserId <$> nextRandom
+  hashedPassword <- traverse hashPassword $ validateShortPassword =<< lookupFormField MissingField "password" paramMap
+  pure $
+    User userId
+    <$> (validateUsername =<< lookupFormField MissingField "username" paramMap)
+    <*> (validateEmailAddress =<< lookupFormField MissingField "email" paramMap)
+    <*> (validateDisplayName =<< lookupFormField MissingField "displayname" paramMap)
+    <*> hashedPassword
+    <*> pure ts
+    <*> pure ts
```

Comment:
The changes here make more sense if you reference `Web.Form` and `Web.Form.Types`